### PR TITLE
Feat/tree selection cascade

### PR DESCRIPTION
**Pull request by switzerb**
Closes [PATH-602](https://gohypergiant.atlassian.net/browse/PATH-602)

Adds a flag to `useTreeActions` and `useTreeState` to change the behavior of selection to cascade. Previously in the tree, selection was always only relevant for the individual node being selected -- selecting a parent did not also select children.

Note there is a separate PR to improve documentation, etc. The code needed it but I wanted separation of concerns: https://github.com/gohypergiant/standard-toolkit/pull/976. 

This PR adds the optional (defaults to false) ability to cascade selection down to children so that selecting a parent will select all the children as well. It also introduces the idea of indeterminate state to the Tree so that if only a portion of children are selected, the ancestors have indeterminate state.

CleanShot 2026-04-14 at 08 14 01

Includes the specs for the change as well.

## ✅ Pull Request Checklist

- [x] Included link to corresponding [GitHub Issue](https://github.com/gohypergiant/standard-toolkit/issues).
- [x] The commit message follows [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) [extended](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type) guidelines.
- [x] Added/updated unit tests and storybook for this change (for bug fixes / features).
- [ ] Added/updated visual regression tests for this change (for bug fixes / features).
- [x] Added/updated documentation (for bug fixes / features)
- [x] Filled out test instructions.
- [x] Added changeset (for bug fixes / features).

## 📝 Test Instructions

 **Test Plan: Tree Cascade Selection**                                                         
                                                                                            
  Setup: Run Storybook and navigate to Tree → CascadeBasic story                            
                                                                                            
  Core Cascade Behavior                                                                     
                                    
  1. Select parent → children cascade: Click checkbox on "Documents" folder. Verify all     
  child files become selected.   
  2. Deselect parent → children cascade: Uncheck "Documents". Verify all children become    
  unselected.                                                                               
  3. Indeterminate state: Manually select only "report.pdf" under "Documents". Verify
  "Documents" shows indeterminate checkbox (dash/minus).                                    
  4. Child-to-parent state: Individually select all children under "Documents". Verify
  parent automatically becomes fully selected.                                              
                                
  Bulk Actions                                                                              
                                
  5. Select All button: Click "Select All". Verify all nodes become selected.               
  6. Unselect All button: Click "Unselect All". Verify all selections clear.
                                                                                            
  Drag & Drop Integration       
                                                                                            
  7. Navigate to Tree → CascadeWithDragAndDrop story                                        
  8. Cascade with drag: Select "Documents" (all children selected). Drag "report.pdf" to
  "Downloads". Verify "Documents" becomes indeterminate.                                    
                                 
  Edge Cases                                                                                
                                
  9. Deep hierarchy: Test selection cascade works through 3+ levels of nesting.             
  10. Disabled nodes: If any disabled nodes exist, verify they're skipped during cascade
  selection.                                                                                
                     
  ---Expected: All scenarios pass without console errors or visual glitches. 

## ❓ Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## 🤖 AI Usage

- [x] Added corresponding label (ai / human) to PR:

If ai was used, select all that apply:
- [x] Ideation / brainstorming
- [x] Documentation
- [x] Testing
- [x] Implementation

### CODE REVIEW

File: packages/design-toolkit/src/hooks/use-tree/actions/index.ts [hunk 1]
```diff
@@ -99,22 +143,84 @@ export function useTreeActions<T>({
...
+      }
+    }
+
+    // Perform move
     for (const key of keys) {
       cache.moveNode(target, key, 0);
     }
 
+    // Sync cascade state (target IS the new parent for moveInto)
+    cache.syncParentsAfterMove(oldParents, target);
```
Comment (colehart) [inline packages/design-toolkit/src/hooks/use-tree/actions/index.ts]: this pattern is repeated in three places. you could DRY this up with a helper

```ts
  /** CASCADE HELPERS **/

  /**
   * Collects the parent keys of all nodes in the given set, before a move.
   */
  function collectOldParents(keys: Set<Key>): Set<Key> {
    const oldParents = new Set<Key>();
    for (const key of keys) {
      const node = cache.getNode(key);
      if (node.parentKey) {
        oldParents.add(node.parentKey);
      }
    }
    return oldParents;
  }
```

and then call that inside `moveAfter`, `moveBefore` and `moveInto`

File: packages/design-toolkit/src/hooks/use-tree/actions/cascade.test.ts [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+      expect(parent?.isIndeterminate).toBe(true);
+
+      // Then unselect all
+      updated = result.current.unselectAll();
+      parent = updated.find((n) => n.key === 'parent');
+
+      expect(parent?.isIndeterminate).toBe(false);
+      expect(parent?.isSelected).toBe(false);
+    });
+  });
```
Comment (colehart) [inline packages/design-toolkit/src/hooks/use-tree/actions/cascade.test.ts:311]: could add tests here for indeterminate with drag and drop:

```suggestion
 describe('Move with cascade (drag-and-drop)', () => {
    const twoParentTree: TreeNode<unknown>[] = [
      {
        key: 'parent1',
        label: 'Parent 1',
        children: [{ key: 'child1', label: 'Child 1' }],
      },
      {
        key: 'parent2',
        label: 'Parent 2',
        children: [{ key: 'child2', label: 'Child 2' }],
      },
    ];

    it('should clear indeterminate state when selected child is moved out via moveAfter', () => {
      const { result } = renderHook(() =>
        useTreeActions({ nodes: simpleTree, selectionCascade: true }),
      );

      // Select child1 only — parent becomes indeterminate
      result.current.onSelectionChange(new Set(['child1']));

      // Move child1 to root level (after parent)
      const updated = result.current.moveAfter('parent', new Set(['child1']));

      const parent = updated.find((n) => n.key === 'parent');

      // parent now only has child2 (unselected), no longer indeterminate
      expect(parent?.isIndeterminate).toBe(false);
      expect(parent?.isSelected).toBe(false);
    });

    it('should clear indeterminate state when selected child is moved out via moveBefore', () => {
      const { result } = renderHook(() =>
        useTreeActions({ nodes: simpleTree, selectionCascade: true }),
      );

      // Select child1 only — parent becomes indeterminate
      result.current.onSelectionChange(new Set(['child1']));

      // Move child1 to root level (before parent)
      const updated = result.current.moveBefore('parent', new Set(['child1']));

      const parent = updated.find((n) => n.key === 'parent');

      // parent now only has child2 (unselected), no longer indeterminate
      expect(parent?.isIndeterminate).toBe(false);
      expect(parent?.isSelected).toBe(false);
    });

    it('should update new parent to indeterminate when a selected child is moved in via moveInto', () => {
      const { result } = renderHook(() =>
        useTreeActions({ nodes: twoParentTree, selectionCascade: true }),
      );

      // Select child1 — parent1 becomes selected (its only selectable child)
      result.current.onSelectionChange(new Set(['child1']));

      // Move child1 into parent2 (which has unselected child2)
      const updated = result.current.moveInto('parent2', new Set(['child1']));

      const parent2 = updated.find((n) => n.key === 'parent2');

      // parent2 now has child1 (selected) and child2 (unselected) → indeterminate
      expect(parent2?.isIndeterminate).toBe(true);
    });
  });
```

Comment (switzerb) [reply packages/design-toolkit/src/hooks/use-tree/actions/cascade.test.ts:311]: I added these tests to the audit branch.

**Review by colehart — APPROVED** [review]
Awesome job!
